File: src/core/ports/outbound/transport.ts
import type { TlsDialOptions, TlsListenOptions } from '../../domain/session/config.js';

export interface Conn {
  readonly id: string;
  readonly remoteAddress: string;
  write(chunk: Buffer): boolean;
  onData(cb: (chunk: Buffer) => void): void;
  onError(cb: (err: Error) => void): void;
  onClose(cb: () => void): void;
  onDrain(cb: () => void): void;
  close(): void;
}

export interface Listener {
  readonly address: string;
  onConnection(cb: (conn: Conn) => void): void;
  onError(cb: (err: Error) => void): void;
  close(): Promise<void>;
}

export interface TransportPort {
  listen(address: string, tls?: TlsListenOptions): Promise<Listener>;
  dial(address: string, tls?: TlsDialOptions): Promise<Conn>;
}

File: internal/port/transport.go
package port

type Conn interface {
	ID() string
	RemoteAddress() string
	Write(chunk []byte) bool
	OnData(cb func(chunk []byte))
	OnError(cb func(err error))
	OnClose(cb func())
	OnDrain(cb func())
	Close()
}

type Listener interface {
	Address() string
	OnConnection(cb func(conn Conn))
	OnError(cb func(err error))
	Close() error
}

type TransportPort interface {
	Listen(address string, tls *TlsListenOptions) (Listener, error)
	Dial(address string, tls *TlsDialOptions) (Conn, error)
}
